tidy3d/components/simulation.py

```diff
@@ -899,6 +899,70 @@ def _validate_2d_geometry_has_2d_medium(cls, val, values):
 
         return val
 
+    @pydantic.validator("structures", always=True)
+    def _validate_advanced_material_intersections(cls, val, values):
+        """Check for unsupported intersections of advanced materials."""
+        structures = val
+        advanced_indices = []
+        advanced_types = []
+        # find and categorize the advanced materials
+        for i, structure in enumerate(structures):
+            if isinstance(structure.medium, AbstractCustomMedium):
+                advanced_indices.append(i)
+                advanced_types.append("custom")
+            elif isinstance(structure.medium, FullyAnisotropicMedium):
+                advanced_indices.append(i)
+                advanced_types.append("fully_anisotropic")
+            elif structure.medium.nonlinear_spec is not None:
+                advanced_indices.append(i)
+                advanced_types.append("nonlinear")
+            elif structure.medium.time_modulated:
+                advanced_indices.append(i)
+                advanced_types.append("time_modulated")
+
+        # check all pairs for possible conflicts
+        # (media that aren't allowed to intersect)
+        conflict_inds = []
+        conflict_types = []
+        for i in range(len(advanced_indices)):
```

Comment:
probably easier to just iterate over
```py
for ind1, type1 in zip(advanced_indices, advanced_types):
```
and same below

Reply:
I think the reason I did it this way is to start the inner for loop at j = i+1, so I don't check each pair twice

Reply:
If we go with your suggestion for adding the medium.is_compatible(other_medium) function, then that breaks the symmetry a bit and we might end up checking each pair twice. Which is ok, just some redundancy which could actually help prevent errors

Reply:
yea, in my code snippet for the new suggestion, I also start the loop at i+1 in the validator , so that should resolve it?

Reply:
ah, right.